Vectorized TWIP training

AugustDG opened the pull request:
TWIP training should be vectorized and properly functioning. This PR does not include a fully trained model (or even optimal parameters).

Closes #7 
Closes #4 
Closes #8

AugustDG committed:
wip: beginning of new training environment

AugustDG committed:
chore: removed unecessary formatting settings

AugustDG committed:
feat: import URDF works

AugustDG committed:
feat: successfully imported working robot

AugustDG committed:
feat: separated base functionality into their respective classes

started to comment why the order of some things matter (specially
imports)

AugustDG committed:
chore: renamed "tasks" -> "core"

AugustDG committed:
feat: added IMU sensor to twip

AugustDG committed:
feat: created somewhat of a base task class

pero que sabes, todo es una m***** (action space?)

AugustDG committed:
chore: switched to Python API for ground plane spawning

+ left a helpful comment for ROS implementation in the future

AugustDG committed:
feat: completed implementation of IVecEnv

AugustDG committed:
feat: hooked up env & agent to task

simulation still works, next step is hooking up to the runner

AugustDG committed:
wip: added base rl_games runner

TODO:
- need to load the rl_games config correctly (with Hydra)
- test the BaseTask implementation
- simplify & clarify the BaseTask (& its derivatives)'s implementations

AugustDG committed:
feat: added config file & basic reading

AugustDG committed:
chore: ignore new links to isaac sim

AugustDG committed:
fix: eliminated errors from loading the config

need to work on making rl_games understand what we give it (by giving it
things that make sense)

AugustDG committed:
feat: cleaned up observations from twip_agent

AugustDG committed:
feat: further progressed on observations for a self-balancing twip

AugustDG committed:
fix: simulation works again

AugustDG committed:
feat: first, somewhat usable, training configuration & task!

AugustDG committed:
feat: updated balancing config

AugustDG committed:
chore: cleaned up project

also exposed more settings in twip.py

AugustDG committed:
wip: vectorizing envs

AugustDG committed:
feat: added multiple envs

AugustDG committed:
feat: made ArticulationView work

Horisofine committed:
Fixed _isaac_sim directory issue

AugustDG committed:
chore: cleaned & adjusted GPU memory allocation values

AugustDG committed:
feat: added cli args & completed refactor

AugustDG committed:
feat: Implemented the RobotView as well as the Replicator

AugustDG committed:
Merge branch 'RobotView' into new-training

AugustDG committed:
Revert "feat: Implemented the RobotView as well as the Replicator"

This reverts commit 0cca87a561b90214523ffcd33c0589e3b4edb243.

AugustDG committed:
Merge pull request #1 from Sim2Capstone/new-training

Merging the refactor changes into my branch

AugustDG committed:
feat: implemented vectorized imu reading

AugustDG committed:
feat: Introduced the replicator and changed the backend

AugustDG committed:
Merge remote-tracking branch 'origin/vectorized-training' into RobotView

AugustDG committed:
feat: Refactored the world setup

Shamiivan commented on core/twip/assets/twip.urdf:
What is the effect of adding stiffness?

AugustDG replied:
If I recall correctly, stiffness should be at 0 for velocity control

Shamiivan reviewed (dismissed):
Looks good to me!

AugustDG commented:
Yayyy! Still in draft tho, so I'll ask for another one then ;)

AugustDG committed:
chore: added an "experience" file to config isaac sim & its extensions

AugustDG committed:
feat: completed vectorized training setup

it's not perfect (imu sensor needs to be manually done), but it's a solid start!

AugustDG committed:
chore: cleaned up some things

AugustDG committed:
feat: added alternative IMU sensor

also tweaked some training parameters

AugustDG committed:
feat: adjusted reward calculation

we also display more info

AugustDG committed:
feat: added episodic resets

AugustDG committed:
feat: added export feature to twip

Shamiivan commented on configs/twip.yaml:
```diff
@@ -0,0 +1,78 @@
+    mlp:
+      units: [256, 128, 64]
+      activation: elu
+      d2rl: False 
```
Any reason this is false?

AugustDG replied:
Mostly because it was false in the original config & I haven't seen people use it. But definitely something that we need to test!

Shamiivan commented on core/envs/generic_env.py:
```diff
@@ -0,0 +1,179 @@
+
+        self.world.reset()
+
+        self.imu = IMU(
```
Is there a way to add noise in the imu values? Not sure if this one but it seems like an ideal IMU. Real ones have noise.

AugustDG replied:
Definitely! I'll make an issue for it :)

Shamiivan commented on core/twip/generic_task.py:
```diff
@@ -0,0 +1,216 @@
+        twip_roll = roll_from_quat(twip_imu_obs[:, 6:10])
+
+        self.obs_buf[:, 0] = twip_roll
+        self.obs_buf[:, 1] = twip_imu_obs[:, 5]  # angular velocity on z-axis
```
Is this gravity free acceleration or gravity taken into account? 
To clarify, when the sensor is stationary, the imu will typically read [0, 0, 1g].
Gravity free acceleration doesn't account for, well, gravity he imu will typically read [0, 0, 0]when stationary. 
Need to know to know which values should be fed into the model

Shamiivan commented on checkpoints/.hydra/config.yaml:
What is the difference between this files and configs/twip.yaml. Why are some of the values different? for example learning_rate, score to win, etc?

AugustDG replied:
I should remove this, good catch! It was how the original system worked (with Hydra), where it would make a copy of the config automatically at every run, but we don't use Hydra anymore.

Shamiivan commented on core/sensors/imu/imu.py:
can you clarify the reference frame?

AugustDG replied:
Done! :)

Shamiivan commented on core/sensors/__init__.py:
Needed? I mean it's an empty file. Any plans for future use?

AugustDG replied:
Yeah, it's necessary for Python to detect modules correctly (dumb, I know)

Shamiivan reviewed (commented):
Not sure about the empty files.
Other than that looks good to me!

AugustDG committed:
docs(imu): modified docs to specify the reference frame

AugustDG committed:
Delete checkpoints/.hydra/hydra.yaml

AugustDG committed:
Delete checkpoints/.hydra/config.yaml

AugustDG committed:
Delete checkpoints/.hydra/overrides.yaml

AugustDG committed:
chore: removed checkpoints folder
